File: SourceCppProject/sources/Parsers/LinesMainProcessor.cpp
#include "LinesMainProcessor.hpp"

#include <algorithm>
#include <functional>
#include <iterator>
#include <utility>
#include <vector>


const LinesStorageMap LinesMainProcessor::extractCommonContent(const LinesStorageMap& first_storage, const LinesStorageMap& second_storage)
{
    LinesStorageMap common_content {};

    const auto first_map = first_storage.getContent();
    const auto second_map = second_storage.getContent();

    for (const std::pair<const std::size_t, std::string>& first_pair : first_map)
    {
        for (const std::pair<const std::size_t, std::string>& second_pair : second_map)
        {
            if (first_pair.first == second_pair.first)
            {
                common_content.putContent(LineInfo(first_pair));
            }
        }
    }

    return common_content;
}

const LinesStorageSet LinesMainProcessor::extractCommonContent(const LinesStorageSet& first_storage, const LinesStorageSet& second_storage)
{
    std::set<std::string> extractedCommonLines;

    std::insert_iterator<std::set<std::string>> inserter = std::inserter(extractedCommonLines, extractedCommonLines.end());

    auto first_src_begin = first_storage.getContent().begin();
    auto first_src_end = first_storage.getContent().end();

    auto second_src_begin = second_storage.getContent().begin();
    auto second_src_end = second_storage.getContent().end();

    std::set_intersection(first_src_begin, first_src_end, second_src_begin, second_src_end, inserter);

    return LinesStorageSet(extractedCommonLines);
}


const LinesStorageMap LinesMainProcessor::extractUniqueContent(const LinesStorageMap& target_storage, const LinesStorageMap& compared_storage)
{
    LinesStorageMap unique_content_map(target_storage.getName());

    const auto target_content_map = target_storage.getContent();
    const auto compared_content_map = compared_storage.getContent();

    for (const std::pair<std::size_t, std::string>& pair : target_content_map)
    {
        if (compared_content_map.count(pair.first) == 0)
        {
            unique_content_map.putContent(LineInfo(pair));
        }
    }

    return unique_content_map;
}

const LinesStorageSet LinesMainProcessor::extractUniqueContent(const LinesStorageSet& target_storage, const LinesStorageSet& compared_storage)
{
    LinesStorageSet extractedUniqueLines(target_storage.getName());

    const std::set<std::string> target_content = target_storage.getContent();
    const std::set<std::string> compared_content = compared_storage.getContent();

    for (const std::string& target_element : target_content)
    {
        if (isUniqueLineInSource(target_element, compared_content))
        {
            extractedUniqueLines.putContent(target_element);
        }
    }

    return extractedUniqueLines;
}

const bool LinesMainProcessor::isUniqueLineInSource(const std::string& unique_object, const std::set<std::string>& source)
{
    if (source.count(unique_object))
    {
        return false;
    }

    return true;
}

File: TextLinesComparing.Library/LinesMainProcessor.cs
namespace TextLinesComparing.Library;

public static class LinesMainProcessor
{
    public static LinesStorageMap ExtractCommonContent(LinesStorageMap first_storage, LinesStorageMap second_storage)
    {
        LinesStorageMap common_content = new();

        Dictionary<int, string> first_map = first_storage.Content;
        Dictionary<int, string> second_map = second_storage.Content;

        foreach (KeyValuePair<int, string> first_pair in first_map)
        {
            foreach (KeyValuePair<int, string> second_pair in second_map)
            {
                if (first_pair.Key == second_pair.Key)
                {
                    common_content.PutContent(new LineInfo(first_pair));
                }
            }
        }

        return common_content;
    }

    public static LinesStorageSet ExtractCommonContent(LinesStorageSet first_storage, LinesStorageSet second_storage)
    {
        SortedSet<string> first_content = first_storage.Content;
        SortedSet<string> second_content = second_storage.Content;
        IEnumerable<string> extractedCommonLines = first_content.Intersect(second_content);
        return new LinesStorageSet(extractedCommonLines);
    }


    public static LinesStorageMap ExtractUniqueContent(LinesStorageMap target_storage, LinesStorageMap compared_storage)
    {
        LinesStorageMap unique_content_map = new()
        {
            Name = target_storage.Name
        };

        Dictionary<int, string> target_content_map = target_storage.Content;
        Dictionary<int, string> compared_content_map = compared_storage.Content;

        foreach (KeyValuePair<int, string> pair in target_content_map)
        {
            if (compared_content_map.ContainsKey(pair.Key) is false)
            {
                unique_content_map.PutContent(new LineInfo(pair));
            }
        }

        return unique_content_map;
    }

    public static LinesStorageSet ExtractUniqueContent(LinesStorageSet target_storage, LinesStorageSet compared_storage)
    {
        LinesStorageSet extractedUniqueLines = new()
        {
            Name = target_storage.Name
        };

        SortedSet<string> target_content = target_storage.Content;
        SortedSet<string> compared_content = compared_storage.Content;

        foreach (var target_element in target_content)
        {
            if (IsUniqueLineInSource(target_element, compared_content))
            {
                extractedUniqueLines.PutContent(target_element);
            }
        }

        return extractedUniqueLines;
    }


    private static bool IsUniqueLineInSource(string unique_object, SortedSet<string> source)
    {
        if (source.Contains(unique_object))
        {
            return false;
        }

        return true;
    }
}
